src/pip/_internal/resolution/resolvelib/base.py

```diff
@@ -20,24 +20,26 @@ def format_name(project, extras):
 
 
 class Constraint:
-    def __init__(self, specifier, hashes):
-        # type: (SpecifierSet, Hashes) -> None
+    def __init__(self, specifier, hashes, links):
+        # type: (SpecifierSet, Hashes, Tuple[Link, ...]) -> None
```

Comment:
I’d prefer something like `Collection[Link]` to make it clearer this is only a bag of links; the fact it’s implemented as a tuple (or list, set etc.) doen’t really matter.